src/Checkout/Helpers/ReserveStock.php

```diff
@@ -0,0 +1,176 @@
+	 *
+	 * @param \WC_Order $order Order object.
+	 * @param int       $minutes How long to reserve stock in minutes. Defaults to woocommerce_hold_stock_minutes.
+	 */
+	public function reserve_stock_for_order( \WC_Order $order, $minutes = 0 ) {
+		$minutes = $minutes ? $minutes : (int) get_option( 'woocommerce_hold_stock_minutes', 60 );
+
+		if ( ! $minutes ) {
+			return;
+		}
+
+		try {
+			$items = array_filter(
+				$order->get_items(),
+				function( $item ) {
+					return $item->is_type( 'line_item' ) && $item->get_product() instanceof \WC_Product && $item->get_quantity() > 0;
+				}
+			);
+			$rows  = array();
+
+			foreach ( $items as $item ) {
+				$product = $item->get_product();
+
+				if ( ! $product->is_in_stock() ) {
+					throw new ReserveStockException(
+						'product_out_of_stock',
+						sprintf(
+							/* translators: %s: product name */
+							__( '%s is out of stock and cannot be purchased.', 'woocommerce' ),
+							$product->get_name()
+						),
+						403
+					);
+				}
+
+				// If stock management is off, no need to reserve any stock here.
+				if ( ! $product->managing_stock() || $product->backorders_allowed() ) {
+					continue;
+				}
+
+				$managed_by_id          = $product->get_stock_managed_by_id();
+				$rows[ $managed_by_id ] = isset( $rows[ $managed_by_id ] ) ? $rows[ $managed_by_id ] + $item->get_quantity() : $item->get_quantity();
+			}
+
+			if ( ! empty( $rows ) ) {
+				foreach ( $rows as $product_id => $quantity ) {
+					$this->reserve_stock_for_product( $product_id, $quantity, $order, $minutes );
+				}
+			}
+		} catch ( ReserveStockException $e ) {
+			$this->release_stock_for_order( $order );
+			throw $e;
+		}
+	}
+
+	/**
+	 * Release a temporary hold on stock for an order.
+	 *
+	 * @param \WC_Order $order Order object.
+	 */
+	public function release_stock_for_order( \WC_Order $order ) {
+		global $wpdb;
+
+		$wpdb->delete(
+			$wpdb->wc_reserved_stock,
+			array(
+				'order_id' => $order->get_id(),
+			)
+		);
+	}
+
+	/**
+	 * Reserve stock for a product by inserting rows into the DB.
+	 *
+	 * @throws ReserveStockException If a row cannot be inserted.
+	 *
+	 * @param int       $product_id Product ID which is having stock reserved.
+	 * @param int       $stock_quantity Stock amount to reserve.
+	 * @param \WC_Order $order Order object which contains the product.
+	 * @param int       $minutes How long to reserve stock in minutes.
+	 */
+	private function reserve_stock_for_product( $product_id, $stock_quantity, \WC_Order $order, $minutes ) {
+		global $wpdb;
+
+		$product_data_store       = \WC_Data_Store::load( 'product' );
+		$query_for_stock          = $product_data_store->get_query_for_stock( $product_id );
+		$query_for_reserved_stock = $this->get_query_for_reserved_stock( $product_id, $order->get_id() );
+
+		// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared, WordPress.DB.PreparedSQL.NotPrepared
+		$result = $wpdb->query(
```

Comment:
perhaps retry 2 3 times in case of deadlocks?

Reply:
I don't think we need a retry mechanism yet because afaik it will wait for the locks, if not user can try again right?

Reply:
if there is a deadlock mysql will revert the query which has done less work by default, so if there is a lock we would want to retry.

Reply:
@vedanshujain @claudiosanches  is this a blocker or something to add in future? The original PR I based this upon had no retry mechanism, and neither does the coupon locking system (hold_applied_coupons) currently in core.

Reply:
I don't think is a blocker, as you said our current code there's no retry mechanism.
For now seems to be just fine, we could include one in the future, or if you want to just prepare it or allow by filters.

Reply:
@mikejolley @claudiosanches  We do have retry system in core currently for coupons see https://github.com/woocommerce/woocommerce/blob/master/includes/data-stores/class-wc-coupon-data-store-cpt.php#L641. 
Note that based on my testing, this query is faster and locks fewer rows than coupons, however, I'd still recommend that we do have retry mechanism.

Reply:
@vedanshujain I mean we don't have one for stock, but I don't anything against including one here.